src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,358 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Code to migrate from a non-Timely-friendly remap shard schema to a
+//! Timely-friendly one.
+use std::sync::Arc;
+
+use anyhow::Context;
+use differential_dataflow::consolidation;
+use differential_dataflow::lattice::Lattice;
+use itertools::Itertools;
+use mz_ore::now::EpochMillis;
+use mz_persist_types::Codec64;
+use timely::order::{PartialOrder, TotalOrder};
+use timely::progress::frontier::{Antichain, MutableAntichain};
+use timely::progress::Timestamp;
+use tokio::sync::Mutex;
+use tracing::info;
+
+use mz_expr::PartitionId;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::read::ReadHandle;
+use mz_persist_client::{PersistClient, ShardId, Upper};
+use mz_repr::{Datum, Diff, GlobalId, TimestampManipulation};
+use mz_timely_util::order::Partitioned;
+
+use crate::types::sources::data::SourceData;
+use crate::types::sources::MzOffset;
+use crate::util::antichain::MutableOffsetAntichain;
+use crate::util::remap_handle::RemapHandle;
+
+use super::CollectionMetadata;
+
+/// A struct capable of performing remap shard migrations. See comment on
+/// [`RemapHandleMigrator::migrate`].
+pub struct RemapHandleMigrator<
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
+    persist_client: PersistClient,
+    current_shard_read_handle: ReadHandle<SourceData, (), T, Diff>,
+    as_of: Antichain<T>,
+    write_shard: ShardId,
+    partitioned_source: bool,
+}
+
+impl<T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation>
+    RemapHandleMigrator<T>
+{
+    /// Prior to this commit, remap shards contained data of schema (Int32?,
+    /// Uint64). Int32 was present only for Kafka sources where it represented
+    /// the Kafka partition.
+    ///
+    /// This schema introduced an impedance mismatch between the remap data and
+    /// native Timely types; tl;dr is that an empty remap shard, if taken
+    /// directly to Timely time types, produces an empty Antichain (i.e. the
+    /// shard is closed from further modification), when in reality we wanted
+    /// it to express the minimum Antichain (i.e. any future input is valid).
+    ///
+    /// TO resolve this, we designed a migration to instead use the schema
+    /// (Range<Int32>?, Uint64). Range<Int32> still represents Kafka
+    /// partitions, but instead is meant to represent seen Kafka partitions, as
+    /// well as the range of those still unseen, e.g.:
+    ///
+    /// (Bottom, 0) -> 0 // trivially empty sub-zero elements in domain
+    /// [0, 1) -> 12 // partition 0 offset
+    /// [1, Top] -> 0 // all unseen partitions
```

Comment:
Discrete-valued ranges (e.g. int32, uint32) rewrite exclusive lower bounds as inclusive bounds + whatever their step value is; in this case + 1. Wrote this in the canonical form because it's necessary for equality to work; however, we could now push `(0, Top)` in and it would be fine; just will never appear the user that way.

Seeing this, realizing I should have written it `[1, Top)` so will make that change.